jgrapht-core/src/main/java/org/jgrapht/alg/CyclicTransitiveReduction.java

```diff
@@ -0,0 +1,285 @@
+        cycle = new ArrayList<>(scComponent.vertexSet());
+        scComponent.removeAllEdges(sccEdgesCopy);
+        for (int i = 0; i < cycle.size(); i++) {
+          V sourceVertex = cycle.get(i);
+          V targetVertex = cycle.get((i + 1) % cycle.size());
+          // Add edge to original graph first, otherwise there will be an exception when trying to add it to a subgraph
+          // of the condensation
+          directedGraph.addEdge(sourceVertex, targetVertex);
+          scComponent.addEdge(sourceVertex, targetVertex);
+        }
+      }
+      else {
+        GraphPath<V, E> graphPath = new DirectedHamiltonianCycle<V, E>().getTour(scComponent);
+        assert graphPath != null;
+        cycle = graphPath.getVertexList();
+        // First and last vertices are identical -> remove one of them
+        cycle.remove(0);
+
+        for (E edge : sccEdgesCopy) {
+          int sourceIndex = cycle.indexOf(scComponent.getEdgeSource(edge));
+          assert sourceIndex >= 0;
+          int targetIndex = cycle.indexOf(scComponent.getEdgeTarget(edge));
+          assert targetIndex >= 0;
+          int indexDelta = Math.abs(sourceIndex - targetIndex);
+          if (indexDelta != 1 && indexDelta + 1 != cycle.size())
+            scComponent.removeEdge(edge);
+        }
+      }
+    }
+  }
+
+  /**
+   * Re-expand a condensed graph previously created by {@link #condenseGraph(Graph)} after possible clean-ups by
+   * {@link #pruneCondensedGraph(Graph)}) back into the original directed graph, also cleaning up  redundant edges
+   * between strongly connected components during the process.
+   *
+   * @param condensedGraph condensed graph to be re-expanded
+   * @param directedGraph target graph to be expanded into
+   */
+  protected void expandCondensedGraph(Graph<Graph<V, E>, DefaultEdge> condensedGraph, Graph<V, E> directedGraph) {
+    // Per condensed graph edge, find one existing edge in the original graph to represent
+    // the corresponding connection between two SCCs
+    Set<E> sccLinks = new HashSet<>();
+    condensedGraph.edgeSet()
+      .forEach(edge -> {
+        Set<V> sourceVertices = condensedGraph.getEdgeSource(edge).vertexSet();
+        Set<V> targetVertices = condensedGraph.getEdgeTarget(edge).vertexSet();
+        E sccLink = allowSyntheticEdges
+          ? createSyntheticSCCLink(sourceVertices, targetVertices)
+          : findSCCLink(sourceVertices, targetVertices);
+        assert sccLink != null;
+        sccLinks.add(sccLink);
+      });
+
+    // Remove all edges from the original graph except for the SCC links
+    new HashSet<E>(directedGraph.edgeSet()).stream()
+      .filter(edge -> !sccLinks.contains(edge))
+      .forEach(directedGraph::removeEdge);
+
+    // Re-add all edges within SCCs
+    condensedGraph.vertexSet().forEach(scComponent ->
+      scComponent.edgeSet().forEach(edge ->
+        directedGraph.addEdge(scComponent.getEdgeSource(edge), scComponent.getEdgeTarget(edge))
```

Comment:
For the non-synthetic edge case, I think we want to preserve the original edge objects (since they may carry additional information) rather than recreating them.

Reply:
We can do that, but it probably makes everything slower.

Listen, would you like to get commit rights in my fork or give me commit rights in a branch and we work on it together? I do not feel so inclined to satisfy all your requirements in a 50-iteration process, having to get to know the whole code base. I would appreciate your support. I have no problem with you first accepting my code into the product as long as it is not faulty, because it adds value. Then feel free to improve and refactor to your heart's content.

Reply:
Well, we're perhaps notorious for our 50-iteration process, and that's unlikely to change :)

Contributors put in their time developing code, and committers put in their time reviewing it to make sure it fits in well with the library, meets quality standards, etc.  Think of us committers as more like editors/publishers rather than as coauthors, and the relationship may work well.

A contributor who has a strong motivation for getting a change included will work with us to make that happen.  In other cases, an abandoned PR may eventually get picked up by someone else and completed.  Either way is fine with us; we all have our own priorities.

In this specific case:  yes, the edge may have other properties (e.g. weights), which is why I was thinking we'd want to preserve it.

Reply:
As you can see, I already implemented it and it was easier than I thought because I discovered the overloaded `addEdge` method which gets an additional edge instance as the 3rd parameter.